tpu-client/src/tpu_client.rs

```diff
@@ -95,6 +105,66 @@ where
         self.invoke(self.tpu_client.try_send_transaction(transaction))
     }
 
+    /// Serialize and send transaction to the current and upcoming leader TPUs according to fanout
+    /// size
+    /// Attempts to send and confirm tx "tries" times
+    /// Waits for signature confirmation before returning
+    /// Returns the transaction signature
+    pub fn send_and_confirm_transaction_with_retries<T: Signers + ?Sized>(
+        &self,
+        keypairs: &T,
+        transaction: &mut Transaction,
+        tries: usize,
+        pending_confirmations: usize,
+    ) -> TransportResult<Signature> {
+        for x in 0..tries {
+            let now = Instant::now();
+            let mut num_confirmed = 0;
+            let mut wait_time = MAX_PROCESSING_AGE;
+            // resend the same transaction until the transaction has no chance of succeeding
+            let wire_transaction =
```

Comment:
maybe move serialization outside of the loop?

Reply:
as is I can't move it out of the loop because at the end of the loop (if the transaction fails to land), the transaction is modified to contain the latest blockhash to send again. See: https://github.com/gregcusack/solana/blob/dd7ad734635ae427916706438c3c0013165b1afe/tpu-client/src/tpu_client.rs#L157-L159